library/core/src/result.rs

```diff
@@ -1023,6 +1023,80 @@ impl<T, E> Result<T, E> {
+    /// that the environment variable was not set when it should have been:
+    ///
+    /// ```should_panic
+    /// let path = std::env::var("IMPORTANT_PATH")
+    ///     .expect("env variable `IMPORTANT_PATH` is not set");
+    /// ```
+    ///
+    /// In the "expect as precondition" style, we would instead describe the
+    /// reason we _expect_ the `Result` should be `Ok`. With this style we would
+    /// prefer to write:
+    ///
+    /// ```should_panic
+    /// let path = std::env::var("IMPORTANT_PATH")
+    ///     .expect("env variable `IMPORTANT_PATH` should be set by `wrapper_script.sh`");
+    /// ```
+    ///
+    /// The "expect as error message" style does not work as well with the
+    /// default output of the std panic hooks, and often ends up repeating
+    /// information that is already communicated by the source error being
+    /// unwrapped:
+    ///
+    /// ```text
+    /// thread 'main' panicked at 'env variable `IMPORTANT_PATH` is not set: NotPresent', src/main.rs:4:6
+    /// ```
+    ///
+    /// In this example we end up mentioning that an env variable is not set,
+    /// followed by our source message that says the env is not present, the
+    /// only additional information we're communicating is the name of the
+    /// environment variable being checked.
+    ///
+    /// The "expect as precondition" style instead focuses on source code
+    /// readability, making it easier to understand what must have gone wrong in
+    /// situations where panics are being used to represent bugs exclusively.
+    /// Also, by framing our expect in terms of what "SHOULD" have happened to
+    /// prevent the source error, we end up introducing new information that is
+    /// independent from our source error.
+    ///
+    /// ```text
+    /// thread 'main' panicked at 'env variable `IMPORTANT_PATH` should be set by `wrapper_script.sh`: NotPresent', src/main.rs:4:6
+    /// ```
+    ///
+    /// In this example we are communicating not only the name of the
+    /// environment variable that should have been set, but also an explanation
+    /// for why it should have been set, and we let the source error display as
+    /// a clear contradiction to our expectation.
+    ///
+    /// For programs where panics may be user facing, either style works best
+    /// when paired with a custom [panic hook] like the one provided by the CLI
+    /// working group library, [`human-panic`]. This panic hook dumps the panic
+    /// messages to a crash report file while showing users a more friendly
+    /// "Oops, something went wrong!" message with a suggestion to send the
+    /// crash report file back to the developers. Panic messages should be used
+    /// to represent bugs, and the information provided back is context intended
+    /// for the developer, not the user.
+    ///
+    /// [panic hook]: https://doc.rust-lang.org/stable/std/panic/fn.set_hook.html
+    /// [`human-panic`]: https://docs.rs/human-panic
```

Comment:
fix: I don't think we should mention specific external crates in the standard library docs, even if they're good ones by official working groups.  (Maybe there are exceptions for crates that come from `rust-lang` repos, but even then I'd expect it to be incredibly rare.)

If we want to link to something, I think it makes more sense to link to a particular [category](https://crates.io/category_slugs) instead -- that way it's a good link so long as that category is meaningful, rather than including a link to a deprecated crate for 6+ weeks if we get another `failure` situation where the suggested thing changes.  (There's not a category that mentions panic hooks right now, but we could fix that.  And maybe link a keyword for now.)

Reply:
We do already mention the log crate, maybe one other crate, can't remember.

I think it's important to link to the specific kind of panic crate that I'm referring too here, I'd love to create a category for crash reporters or crash dump hook or something like that, but it's more specific than just panic hooks.

Reply:
`log` hits the exception I mentioned though, because it's
![image](https://user-images.githubusercontent.com/18526288/167207144-700eba64-3e6f-4c47-922f-115f8030949e.png)

A rust-cli book mentioning human-panic makes good sense to me, but I don't think `std` should.  Especially since it's cargo metadata is still a person's repo
![image](https://user-images.githubusercontent.com/18526288/167207356-f95baa12-5300-4006-8244-5dee63d0040e.png)

Reply:
I'll buy that, tho I think this is a sign that we should provide something like human-panic that we can point at.